### ISSUE TITLE
Gradient checkpointing broken when used within transformers (following the seq-2-seq tutorial).

### ISSUE THREAD

**Issue by Vbansal21** [pytorch/pytorch#59241]
When torch.utils.checkpoint.checkpoint is used on transformer layer(s) in the model discribed in the seq-2-seq tutorial on pytorch website, it renders the transformer layers almost useless, the model won't converge, and behave similar to the model without transformer layer(s).

changes to code:
`output = self.transformer_encoder(src, src_mask)`
to
`output = torch.utils.checkpoint.checkpoint(self.transformer_encoder,src, src_mask)`
and the model becomes useless.
(all else of the code is same, as discribed in the tutorial)

cc @zhangguanheng66

**Comment by jbschlosser** [comment]
Thanks for the report! Could you give a bit more information on which version of PyTorch you're using, etc. (ideally run the
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py) and report the results here).

**Comment by Vbansal21** [comment]
> Thanks for the report! Could you give a bit more information on which version of PyTorch you're using, etc. (ideally run the
> [environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py) and report the results here).

torch 1.8.1 (all things default to google-colab setup, nothing changed)

**Comment by Vbansal21** [comment]
Closing issue. The problem seems to have been solved (from pytorch library side).
